osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs

```diff
@@ -0,0 +1,110 @@
+    /// </summary>
+    public class ReverseQueue<T> : IEnumerable<T>
+    {
+        /// <summary>
+        /// The number of elements in the <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        public int Count { get; private set; }
+
+        private T[] items;
+        private int capacity;
+        private int start;
+
+        public ReverseQueue(int initialCapacity)
+        {
+            if (initialCapacity <= 0)
+                throw new ArgumentOutOfRangeException(nameof(initialCapacity));
+
+            items = new T[initialCapacity];
+            capacity = initialCapacity;
+            start = 0;
+            Count = 0;
+        }
+
+        /// <summary>
+        /// Retrieves the item at an index in the <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        /// <param name="index">The index of the item to retrieve. The most recently enqueued item is at index 0.</param>
+        public T this[int index]
+        {
+            get
+            {
+                if (index < 0 || index > Count - 1)
+                    throw new ArgumentOutOfRangeException(nameof(index));
+
+                int reverseIndex = Count - 1 - index;
+                return items[(start + reverseIndex) % capacity];
+            }
+        }
+
+        /// <summary>
+        /// Enqueues an item to this <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        /// <param name="item">The item to enqueue.</param>
+        public void Enqueue(T item)
+        {
+            if (Count == capacity)
+            {
+                // Double the buffer size
+                var buffer = new T[capacity * 2];
+
+                // Copy items to new queue
+                for (int i = 0; i < Count; i++)
+                {
+                    buffer[i] = items[(start + i) % capacity];
+                }
+
+                // Replace array with new buffer
+                items = buffer;
+                capacity *= 2;
+                start = 0;
+            }
+
+            items[(start + Count) % capacity] = item;
+            Count++;
+        }
+
+        /// <summary>
+        /// Dequeues an item from the <see cref="ReverseQueue{T}"/> and returns it.
+        /// </summary>
+        /// <returns>The item dequeued from the <see cref="ReverseQueue{T}"/>.</returns>
+        public T Dequeue()
+        {
+            var item = items[start];
+            start = (start + 1) % capacity;
+            Count--;
+            return item;
+        }
+
+        /// <summary>
+        /// Clears the <see cref="ReverseQueue{T}"/> of all items.
+        /// </summary>
+        public void Clear()
+        {
+            start = 0;
+            Count = 0;
+        }
+
+        /// <summary>
+        /// Returns an enumerator which enumerates items in the <see cref="ReverseQueue{T}"/> starting from the most recently enqueued one.
+        /// </summary>
+        public IEnumerator<T> GetEnumerator()
```

Comment:
Is this intended to be used a lot? This will incur an allocation every time unless a custom enumerator is added. Something like this: https://github.com/ppy/osu-framework/blob/07cef87ebbb592e97e1b965b3f4147da03570352/osu.Framework/Graphics/Containers/Container.cs#L458-L480

Reply:
Currently it exists purely to implement the IEnumerable<T> interface, and is only referenced in the test for it.
That said, I do imagine implementations that require a dynamic history will be using the enumerator via foreach loops, so it sounds like a good thing to account for. :+1:

Reply:
Fixed in bfd3d0cce978c77d4bfa7bbe1a15262e4cc23c80